generate_crossword/step_10/lib/model.dart

```diff
@@ -0,0 +1,724 @@
+      return null;
+    }
+
+    final wordCharacters = word.characters;
+    bool overlap = false;
+
+    // Check that the word fits in the crossword.
+    for (final (index, character) in wordCharacters.indexed) {
+      final characterLocation = switch (direction) {
+        Direction.across => location.rightOffset(index),
+        Direction.down => location.downOffset(index),
+      };
+
+      final target = characters[characterLocation];
+      if (target != null) {
+        overlap = true;
+        if (target.character != character) {
+          return null;
+        }
+        if (direction == Direction.across && target.acrossWord != null ||
+            direction == Direction.down && target.downWord != null) {
+          return null;
+        }
+      }
+    }
+
+    // If overlap is required, make sure that the word overlaps with an existing
+    // word. Skip this test if the crossword is empty.
+    if (words.isNotEmpty && !overlap && requireOverlap) {
+      return null;
+    }
+
+    final candidate = rebuild(
+      (b) => b
+        ..words.add(
+          CrosswordWord.word(
+            word: word,
+            direction: direction,
+            location: location,
+          ),
+        ),
+    );
+
+    if (candidate.valid) {
+      return candidate;
+    } else {
+      return null;
+    }
+  }
+
+  /// As a finalize step, fill in the characters map.
+  @BuiltValueHook(finalizeBuilder: true)
+  static void _fillCharacters(CrosswordBuilder b) {
+    b.characters.clear();
+
+    for (final word in b.words.build()) {
+      for (final (idx, character) in word.word.characters.indexed) {
+        switch (word.direction) {
+          case Direction.across:
+            b.characters.updateValue(
+              word.location.rightOffset(idx),
+              (b) => b.rebuild((bInner) => bInner.acrossWord.replace(word)),
+              ifAbsent: () => CrosswordCharacter.character(
+                acrossWord: word,
+                character: character,
+              ),
+            );
+          case Direction.down:
+            b.characters.updateValue(
+              word.location.downOffset(idx),
+              (b) => b.rebuild((bInner) => bInner.downWord.replace(word)),
+              ifAbsent: () => CrosswordCharacter.character(
+                downWord: word,
+                character: character,
+              ),
+            );
+        }
+      }
+    }
+  }
+
+  /// Pretty print a crossword. Generates the character grid, and lists
+  /// the down words and across words sorted by location.
+  String prettyPrintCrossword() {
+    final buffer = StringBuffer();
+    final grid = List.generate(
+      height,
+      (_) => List.generate(
+        width, (_) => '░', // https://www.compart.com/en/unicode/U+2591
+      ),
+    );
+
+    for (final entry in characters.entries) {
+      final location = entry.key;
+      final character = entry.value;
+      grid[location.y][location.x] = character.character;
+    }
```

Comment:
Another syntax to consider:

```suggestion
    for (final MapEntry(key: Location(:x, :y), value: character)
        in characters.entries) {
      grid[y][x] = character.character;
    }
```

or just the following, if you don't want to repeat `Location`:

```suggestion
    for (final MapEntry(key: location, value: character)
        in characters.entries) {
      grid[location.y][location.x] = character.character;
    }
```

Reply:
That's a tad wild.